scripts/Diffraction/isis_powder/pearl.py

```diff
@@ -24,9 +27,11 @@ def __init__(self, **kwargs):
     def focus(self, **kwargs):
         self._switch_long_mode_inst_settings(kwargs.get("long_mode"))
         self._inst_settings.update_attributes(kwargs=kwargs)
-        return self._focus(run_number_string=self._inst_settings.run_number,
-                           do_absorb_corrections=self._inst_settings.absorb_corrections,
-                           do_van_normalisation=self._inst_settings.van_norm)
+        focused_run = self._focus(run_number_string=self._inst_settings.run_number,
```

Comment:
I can see that this makes sense as the additional arguments should only apply for this function call. Could the approach of having to remember to restore the settings afterwards be error prone though? What if `self._focus` throws an exception and the final line doesn't run?

Could we use a decorator or a context manager? For example a context manager would look like:

```python
def focus(self, **kwargs):
    with restore_defaults_post_process():
         self._switch_long_mode_inst_settings(kwargs.get("long_mode"))
         self._inst_settings.update_attributes(kwargs=kwargs)
         return self._focus(run_number_string=self._inst_settings.run_number, ...
```
or a decorator (but I'm not sure how this plays with exceptions)

```python
@restore_defaults_post_process
def focus(self, **kwargs):
     self._switch_long_mode_inst_settings(kwargs.get("long_mode"))
     self._inst_settings.update_attributes(kwargs=kwargs)
     return self._focus(run_number_string=self._inst_settings.run_number, ...
```

Reply:
Ooh yes, I thought there must be a nicer way of doing it